llvm/lib/IR/RepoDefinition.cpp

```diff
@@ -169,29 +169,42 @@ void ModuleHashGenerator::calculateGONumAndGOIMap(const Module &M) {
 #endif
 }
 
-void ModuleHashGenerator::updateDigestUseContributions(
-    MD5 &GOHash, const GOVec &Contributions) {
-  for (const GlobalObject *const G : Contributions) {
-    GOHash.update(GOIMap[G].InitialDigest.Bytes);
-  }
+size_t ModuleHashGenerator::updateDigestUseContributions(
+    const GlobalObject *GO, MD5 &GOHash, const GOVec &Contributions,
+    bool UseRepoDefinitionMD) {
   // GOHash/Module is changed if Contributions is not empty.
   Changed = Changed || !Contributions.empty();
+
+  auto LoopPoint = std::numeric_limits<size_t>::max();
+  for (const GlobalObject *const G : Contributions) {
+    // Although 'contributions' and 'dependencies' respective arrows point are
+    // in different direction,  they can affect each other and form loops.
+    // Consider the following example:
+    // int var = 0;
+    // int Func() {
+    //    var = var + 5;
+    //    return var;
+    //  }
+    // 'Func' is a contribution of 'var' and 'var' is a dependency of 'Func`.
+    // Their hash values are dependent on each other. They can form loops.
+    size_t GDepth;
+    DigestType GDigest;
+    std::tie(GDepth, GDigest) =
+        updateDigestUseDependenciesAndContributions(G, UseRepoDefinitionMD);
+    // A GO which loops back to itself doesn't count as a loop.
+    if (G != GO)
+      LoopPoint = std::min(LoopPoint, GDepth);
+    GOHash.update(GDigest.Bytes);
+  }
+  return LoopPoint;
```

Comment:
We don't record anything in the hash to indicate where the contribution hashes and and the dependency hashes start.

Reply:
Again, if we can do your advice "convert the contributions to dependencies", this problem will  be solved.